From 2f496e9c6c28854bd52dd786b3a80215a429ea7c Mon Sep 17 00:00:00 2001
From: Yanan Yang <yanan.yang@nxp.com>
Date: Thu, 20 Mar 2025 11:24:09 +0900
Subject: [PATCH 20/30] thermal: imx91: bug fix: Temperature read returns
 "Resource temporarily unavailable" for values below 0 degree

This issue occurs because the register DATA0[DATA_VAL] is a 16-bit signed value,
but the temperature was not being processed as such before being converted to a 32-bit value.

The fix ensures that the temperature is correctly calculated as a 16-bit signed value
before the conversion to a 32-bit format. This prevents misinterpretation of negative
temperatures and resolves the "Resource temporarily unavailable" error.

Signed-off-by: Yanan Yang <yanan.yang@nxp.com>
---
 drivers/thermal/imx91_thermal.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/thermal/imx91_thermal.c b/drivers/thermal/imx91_thermal.c
index c8d1594ad359..595a48c65a49 100644
--- a/drivers/thermal/imx91_thermal.c
+++ b/drivers/thermal/imx91_thermal.c
@@ -114,7 +114,7 @@ static int imx91_tmu_get_temp(struct thermal_zone_device *tz, int *temp)
 		return -EAGAIN;
 
 	val = readl_relaxed(tmu->base + DATA0) & 0xffffU;
-	*temp = (int)val * 1000LL / 64LL;
+	*temp = (int)((int16_t)val * 1000LL / 64LL);
 	if (*temp < TMU_TEMP_LOW_LIMIT || *temp > TMU_TEMP_HIGH_LIMIT)
 		return -EAGAIN;
 
-- 
2.34.1

